### DISCUSSION TITLE
OIDCPreservePost not preserving newlines in POST data for textarea input fields

### DISCUSSION THREAD

**Discussion by jansmets** [OpenIDC/mod_auth_openidc#717] [Q&A]
Hi!

I'm trying to preserve POST data.  The form has a textarea field.

The submitted form data is

"test
test
test
test"

During the redirection dance I see 
 mod_auth_openidc_preserve_post_params => {"test":"test%0D%0Atest%0D%0Atest%0D%0Atest%0D%0A"}
which looks good to me. (%0D%0A = CR LF)

After authentication with the idp the browser goes back to the redirect_uri, and then finally the browser does a POST again to the original location. In that request I see "testtesttesttest" - which has the newlines removed.

The Javascript used to replay the POST seems to be dropping that data.

          var input = document.createElement("input");
https://github.com/zmartzone/mod_auth_openidc/blob/master/src/mod_auth_openidc.c#L494

behaves differently and is only valid for input elements, not textarea's. For example:

data="test%0D%0Atest%0D%0Atest%0D%0Atest%0D%0A"

input = document.createElement("input");
input.value = decodeURIComponent(data)
input.value
=> 'testtesttesttest'
area = document.createElement("textarea")
area.value = decodeURIComponent(data)
area.value
=> 'test\ntest\ntest\ntest\n'

I believe the element type also needs to be stored in the json data.

Thank you

**Comment by zandbelt** [comment]
indeed the feature was designed only for simple name/value inputs; extending it in the way you mention would probably work

**Reply by paulQdata** [reply]
To fix the newline problem, It's probably sufficient to set the input type to "hidden" before setting the value.

**Reply by zandbelt** [reply]
I believe it is hidden already: https://github.com/OpenIDC/mod_auth_openidc/blob/master/src/mod_auth_openidc.c#L514

**Reply by paulQdata** [reply]
The order matters. Try loading this html-page for example

> <html>
	<body>
		<input type="hidden" id="testhidden"/>
		<input type="text" id="testtext" />
		<input id="testempty" />
	<script>
	 function setAndTestValue(elementId){
	 	 var testVal = '123\nd234';
		 var el = document.getElementById(elementId);
		 el.value = testVal;
		 console.log('test: ' + elementId);
		 console.log((el.value === testVal));
	 }

	 ['testhidden','testtext', 'testempty'].forEach(setAndTestValue);
	</script>
	</body>
</html>

**Reply by zandbelt** [reply]
ok, would you be able to test this?

**Reply by paulQdata** [reply]
No. I was just trying to check if we should enable OIDCPreservePost. I don't have access to anything like a working test-environment.

**Reply by zandbelt** [reply]
I verified this successfully, thanks for chiming in! https://github.com/OpenIDC/mod_auth_openidc/commit/2b40e4137d3e85a1899d61198bf23127d7afee57
